DM-31785: Improve efficiency of butler.transfer_from

timj opened the pull request:
* Add `datastore.mexists()` and `ButlerURI.mexists()`
* Use parallel queries of object store.
* Cache result of initial check for registry entries and reuse for check inside datastore.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Add basic datastore.mexist() implementation

Does not parallelize existence checks (yet)

timj committed:
Use chunking in datastore.mexist

timj committed:
Use concurrent futures in ButlerURI.mexists

timj committed:
Improve chunking log output in mexists

timj committed:
Adjust some log messages in mexists

With chunking we are getting feedback above reporting how many
were really missing.

timj committed:
Keep track of artifact existence in transfer_from

We check for dataset existence once in Butler.transfer_from
to filter out missing datasets that we don't want to go in
the registry. Then later on we check inside datastore
to see which files exist because composite disassembly
means that we can't know up front whether a file
should be there or not.

timj committed:
Also chunk the datastore.transfer_from existence checks

This only triggers if called without the lookup table.

timj committed:
Ensure that artifact existence is propagated in chained datastore

timj committed:
Add simple test for datastore.mexists

ktlim commented on python/lsst/daf/butler/core/_butlerUri/_butlerUri.py:
```diff
@@ -653,6 +654,43 @@ def exists(self) -> bool:
+        # The max worker count here must not exceed the max connection pools
+        # in whatever system is doing the network connection. The default
+        # is 10 for urllib3.
+        max_workers = 10
```
Seems strange to have a hard-coded constant like this.  Seems like it should be at minimum a class variable or perhaps a kwarg.

timj replied:
Okay. Will move it to be a package constant for now. This will get revisited when I have to implement the parallelized transfer code.

ktlim commented on python/lsst/daf/butler/core/_butlerUri/_butlerUri.py:
```diff
@@ -653,6 +654,43 @@ def exists(self) -> bool:
+                exists = False
+            return (uri, exists)
+
+        future_exists = [exists_executor.submit(wrapper, uri) for uri in uris]
```
Can't this all be replaced by `return dict(exists_executor.map(wrapper, uris))`?

timj replied:
Will look. I'm a newbie on this sort of thing so was mostly using the code that was in bps as a template.

timj replied:
I changed it to use the dict approach recommended in the python docs and simplified it a lot.

ktlim commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1949,21 +2127,57 @@ def transfer_from(self, source_datastore: Datastore, refs: Iterable[DatasetRef],
+                iterable."""
+                it = iter(data)
+                for i in range(0, len(data), chunk_size):
+                    yield tuple(element for element in itertools.islice(it, chunk_size))
```
I don't think you need the `element for element in`; simply converting to `tuple` is enough.

ktlim commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1244,6 +1248,175 @@ def knows(self, ref: DatasetRef) -> bool:
+        """
+        chunk_size = 10_000
+        dataset_existence: Dict[DatasetRef, bool] = {}
+        refs = list(refs)
```
Why two different ways of chunking?  `itertools.islice()` below in `transfer_from()` versus materializing the list and using regular slicing here?

timj replied:
I'm a bit confused because I could have sworn that I had already replied to this comment. I can't find it though.

Some of this is history and some of it is the code (not just this bit) being terrified that it might be called with an `Iterator` and not a `Collection`. In this particular case it is probably best to only create new small lists rather than one big list.

There are probably many cases in the code where if we simply assume up top that it was a List then things would be simpler. The problem is that users tend to pass in either an iterator (as returned by `queryDatasets`) or a set (which is the standard advice for what to do with `queryDatasets` output). Iterator isn't Sized but set is. Set is a Collection but you can't index into it.

timj replied:
I moved `chunk_iterable` to the utilities package and reimplemented it so it would work without needing `len()` to work. Even though it's only 3 lines long I think it is easier to use as `for chunk in chunk_iterable(x)` format but maybe it's one of those cases where:

```python
for chunk in chunk_iterable(data, 4):
    # use chunk
```
vs
```python
while chunk := tuple(itertools.islice(iter(data), 4)):
    # use chunk
```
is not that much harder to use if you can remember it.

ktlim commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1949,21 +2127,57 @@ def transfer_from(self, source_datastore: Datastore, refs: Iterable[DatasetRef],
-                else:
-                    log.warning("Asked to transfer dataset %s but no file artifacts exist for it.",
-                                id_to_ref[missing])
+            def chunk_iterable(data: Collection[Any], chunk_size: int = 10_000) -> Iterator[Iterable[Any]]:
```
`Iterable` or `Collection`?

ktlim commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1244,6 +1248,175 @@ def knows(self, ref: DatasetRef) -> bool:
+            if log.isEnabledFor(VERBOSE):
+                n_results = len(chunk_result)
+                n_checked += n_results
+                n_found = 0
```
`n_found = sum(chunk_result.values())`

timj replied:
Am I allowed to assume that bool will translate as 1 or 0 because there is an implicit `int(True)` involved in this sum?

timj replied:
Seems like I am. Will change.

ktlim commented on python/lsst/daf/butler/core/_butlerUri/_butlerUri.py:
```diff
@@ -653,6 +654,43 @@ def exists(self) -> bool:
+
+        Returns
+        -------
+        existence : `dict` of [`ButlerURI`, `bool`]
```
I think that with proper handling this (and the similar return for `Datastore.mexists()`) can be replaced by a simpler `set of ButlerURI`, but that can be left for later.

timj replied:
I haven't yet understood how one set can do it. I can see how a found set and not-found set can do it.

timj committed:
Add codecov label

timj committed:
Enable terminal output for code coverage and explicit butler module.

The terminal output allows us to compare with codecov.
Since the butler package is installed we need to specify the
lsst.daf.butler module for code coverage.

timj committed:
Use pip install -e

This will stop it being installed directly so will simplify the
coverage report.

timj committed:
Fix flake8 in version.py

timj committed:
Install utils directly from git

timj committed:
Improve test coverage for butler.transfer_from

This takes into account the missing data scenario of
execution butler.

timj committed:
Add change record

timj committed:
Use package constant for max workers

timj committed:
Move import to top of file

It is going to be used by other functions and should never
have been embedded.

timj committed:
Simplify futures logic for mexists()

timj committed:
Add simple ButlerURI.mexists test

timj committed:
Simplify count for True values in dict

Also ensure that we test that the code is triggered.

timj committed:
Add a new utility routine to chunk an iterable into smaller pieces

timj committed:
Unify both places where iterables are chunked

Use the new chunk_iterable utility function.

timj committed:
Use source datastore as factory rather than target

Without this the code starts looking in the wrong datastore.

timj committed:
Add test for transfers of disassembled composites
